### ISSUE TITLE
add image gradient op

### ISSUE THREAD

**Issue by bodokaiser** [tensorflow/tensorflow#13228]
A [paper][1] I am reimplementing recently uses [image gradient loss][2]. Numpy offers [np.gradient][3] to achieve this task i.e. `np.gradient(image, axis=0), np.gradient(image, axis=1)` however tensorflow lacks this feature or at least documentation about how to use `tf.gradients` to get this done.

Therefore I propose to either send a PR where I add a gradient image op which uses fixed 2d convolution, i.e.:

```python
import numpy as np
import scipy as sp
import tensorflow as tf

from skimage import io
from skimage import color

from matplotlib import pyplot as plt

image = color.rgb2gray(io.imread('image.jpg'))
image_rs = image.reshape([1] + list(image.shape) + [1])

xgrad = np.gradient(image, axis=0)
ygrad = np.gradient(image, axis=1)

image_ph = tf.placeholder(tf.float32, image_rs.shape)

x_weight = tf.reshape(tf.constant([-1, 0, +1], tf.float32), [3, 1, 1, 1])
y_weight = tf.reshape(x_weight, [1, 3, 1, 1])

xgrad_ts = tf.nn.conv2d(image_ph, x_weight, [1, 1, 1, 1], 'SAME')
ygrad_ts = tf.nn.conv2d(image_ph, y_weight, [1, 1, 1, 1], 'SAME')

with tf.Session() as sess:
    xgrad2, ygrad2 = sess.run([xgrad_ts, ygrad_ts], feed_dict={image_ph: image_rs})

    print(xgrad2.shape)
    fig, axes = plt.subplots(2, 3)
    axes[0, 0].imshow(image)
    axes[0, 1].imshow(xgrad)
    axes[0, 2].imshow(ygrad)
    axes[1, 0].imshow(image)
    axes[1, 1].imshow(xgrad2[0,:,:,0])
    axes[1, 2].imshow(ygrad2[0,:,:,0])
    plt.show()

```

or have someone update documentation of `tf.gradients`.

For the first I could provide a PR if this is considered interesting for tensorflow and not "too implementation specific".

[1]: https://arxiv.org/abs/1612.05362
[2]: https://en.wikipedia.org/wiki/Image_gradient
[3]: https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.gradient.html

**Comment by aselle** [comment]
You are mixing up spatial (image gradients) which are respect to a change in position and gradients with respect to parameters of the ML model. Using a fixed 2d convolution as you show here is the correct solution. tf.gradients is used only for computing gradients with respect to variables in your model. In this case, creating a convolution that uses a constant to implement a gradient is what you want to do. Creating a helper function to do this may or may not be interesting. @shlens, do you have any thoughts?

**Comment by shlens** [comment]
Agreed with @aselle. Note the spatial gradient is computed in [total_variation](https://www.tensorflow.org/api_docs/python/tf/image/total_variation) and you might be able to employ this function as a prototype for your particular needs.

**Comment by bodokaiser** [comment]
Yes, [total_variation][1] was what I was looking for. Its implementation should also be more efficient then the one I came up [2]. 

[1]: https://www.tensorflow.org/api_docs/python/tf/image/total_variation
[2]: https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/ops/image_ops_impl.py#L1455-L1466

**Comment by aselle** [comment]
Glad it resolved your issue!

**Comment by micklexqg** [comment]
@aselle , I think the total_variation(according to the given code) is for difference but not for gradient. The two are different.
